pyccl/pk2d.py

```diff
@@ -248,13 +253,145 @@ def eval_dlogpk_dlogk(self, k, a, cosmo):
+        if (self.psp.amin < other.psp.amin
+                or self.psp.amax > other.psp.amax):
+            raise ValueError("The 2nd operand has its data defined over a "
+                             "smaller a range than the 1st operand.")
+
+        a_arr_a, lk_arr_a, pk_arr_a = self.get_spline_arrays()
+        a_arr_b, lk_arr_b, pk_arr_b = other.get_spline_arrays()
+        if not (a_arr_a.size == a_arr_b.size and lk_arr_a.size == lk_arr_b.size
+                and np.allclose(a_arr_a, a_arr_b)
+                and np.allclose(lk_arr_a, lk_arr_b)):
+            # Since the power spectrum is evalulated on a smaller support than
+            # where it was defined, no extrapolation is necessary and the
+            # dependence on the cosmology in moot.
+            # CosmologyVanillaLCDM is being imported here instea of the top of
+            # the module due to circular import issues there.
+            from .core import CosmologyVanillaLCDM
+            dummy_cosmo = CosmologyVanillaLCDM()
+            pk_arr_b = np.array([other.eval(k=np.exp(lk_arr_a),
+                                            a=a_,
+                                            cosmo=dummy_cosmo)
+                                 for a_ in a_arr_a])
+        return a_arr_a, lk_arr_a, pk_arr_a, pk_arr_b
+
+    def __add__(self, other):
+        """Adds two Pk2D instances.
+
+        The a and k ranges of the 2nd operand need to be the same or smaller
+        than the 1st operand.
+        The returned Pk2D object uses the same a and k arrays as the first
+        operand.
+        """
+        a_arr_a, lk_arr_a, pk_arr_a, pk_arr_b = \
+            self._check_Pk2D_compatability(other)
+        pk_arr_new = pk_arr_a + pk_arr_b
+
+        logp = np.all(pk_arr_new > 0)
+        if logp:
+            pk_arr_new = np.log(pk_arr_new)
+
+        new = Pk2D(a_arr=a_arr_a, lk_arr=lk_arr_a, pk_arr=pk_arr_new,
+                   is_logp=logp,
+                   extrap_order_lok=self.extrap_order_lok,
+                   extrap_order_hik=self.extrap_order_hik)
+
+        return new
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    def __mul__(self, other):
+        """Multiply two Pk2D instances.
+
+        The a and k ranges of the 2nd operand need to be the same or smaller
+        than the 1st operand.
+        The returned Pk2D object uses the same a and k arrays as the first
+        operand.
+        """
+        if isinstance(other, (float, int)):
+            a_arr_a, lk_arr_a, pk_arr_a = self.get_spline_arrays()
+            pk_arr_new = other * pk_arr_a
+        elif isinstance(other, Pk2D):
+            a_arr_a, lk_arr_a, pk_arr_a, pk_arr_b = \
+                self._check_Pk2D_compatability(other)
+            pk_arr_new = pk_arr_a * pk_arr_b
+        else:
+            raise ValueError("Multiplication of Pk2D is only defined for "
+                             "floats, ints, and Pk2D objects.")
```

Comment:
Would it be useful to do the same for `__add__`?
i.e. make `my_pk + 0.002` a valid call?

Also, would it be useful to accept ready-made 2d arrays (of narrower support maybe) and multiply them without accessing the data from the 2nd operand?
i.e. `Pk * my_2d_array_k_a`

Reply:
Great, thank you!
Agreed on the second point, so I'm now resolving this.